File: 0695-max-area-of-island/0695-max-area-of-island.java
class Solution {
    int[] dr = new int[]{1, -1, 0, 0};
    int[] dc = new int[]{0, 0, 1, -1};
    int count = 0;
    int max = 0;
    public int maxAreaOfIsland(int[][] grid) {
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                if (grid[i][j] == 1) {
                    dfs(grid, i, j);
                    if (count > max) {
                        max = count;
                    }
                    count = 0;
                }
            }
        }

        return max;
    }

    private void dfs(int[][] grid, int row, int col) {
        if (row < 0 || row >= grid.length || col < 0 || col >= grid[0].length || grid[row][col] == 0) {
            return;
        }

        grid[row][col] = 0;
        count++;
        for (int k = 0; k < 4; k++) {
            int newRow = row + dr[k];
            int newCol = col + dc[k];
            dfs(grid, newRow, newCol);
        }
    }
}

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    
    def __init__(self):
        self.res = 0
        self.currMax = 0   

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m = len(grid)
        n = len(grid[0])
        dr = [1, -1, 0, 0]
        dc = [0, 0, 1, -1]

        def dfs(x, y):
            grid[x][y] = 0
            self.currMax += 1

            for k in range(4):
                nx, ny = x + dr[k], y + dc[k]

                if 0 <= nx < m and 0 <= ny < n and grid[nx][ny] == 1:
                    dfs(nx, ny)

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    dfs(i, j)
                    self.res = max(self.res, self.currMax)
                    self.currMax = 0

        return self.res